### ISSUE TITLE
setup:upgrade - Area code not set: Area code must be set before starting a session.

### ISSUE THREAD

**Issue by macsanta** [mageplaza/magento-2-blog#248]
Mageplaza_Blog - Area code not set: Area code must be set before starting a session.
### Preconditions

1. Magento: 2.2.7
2. PHP: 7.1
3. MySQL:  5.7.28
4. Blog v3.1.0

**Comment by Bang512** [comment]
Hi @macsanta,

Thanks for contacting us.
I'm not sure I can understand your problem.
Could you please explain your problem in more detail and send me some screenshots?

With it, I can help you solve this problem in the shortest time.
Thanks!

**Comment by andyexeter** [comment]
Experiencing the same issue upgrading from 3.0.x to 3.1.1. The following error is thrown when running `php bin/magento setup:db-schema:upgrade -vvv`:

```
Module 'Mageplaza_Blog':

                                                                       
  [Magento\Framework\Exception\SessionException]                       
  Area code not set: Area code must be set before starting a session.  
                                                                       

Exception trace:
 () at /var/www/app/vendor/magento/framework/Session/SessionManager.php:175
 Magento\Framework\Session\SessionManager->start() at /var/www/app/generated/code/Magento/Framework/Session/Generic/Interceptor.php:50
 Magento\Framework\Session\Generic\Interceptor->start() at /var/www/app/vendor/magento/framework/Session/SessionManager.php:130
 Magento\Framework\Session\SessionManager->__construct() at /var/www/app/generated/code/Magento/Framework/Session/Generic/Interceptor.php:14
 Magento\Framework\Session\Generic\Interceptor->__construct() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:111
 Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:66
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() at /var/www/app/vendor/magento/framework/ObjectManager/ObjectManager.php:70
 Magento\Framework\ObjectManager\ObjectManager->get() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:144
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:230
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:34
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:59
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() at /var/www/app/vendor/magento/framework/ObjectManager/ObjectManager.php:70
 Magento\Framework\ObjectManager\ObjectManager->get() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:144
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:230
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:34
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:59
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() at /var/www/app/vendor/magento/framework/ObjectManager/ObjectManager.php:70
 Magento\Framework\ObjectManager\ObjectManager->get() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:144
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:230
 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:34
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments() at /var/www/app/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:59
 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() at /var/www/app/vendor/magento/framework/ObjectManager/ObjectManager.php:56
 Magento\Framework\ObjectManager\ObjectManager->create() at /var/www/app/setup/src/Magento/Setup/Model/Installer.php:963
 Magento\Setup\Model\Installer->createSchemaDataHandler() at /var/www/app/setup/src/Magento/Setup/Model/Installer.php:1267
 Magento\Setup\Model\Installer->getSchemaDataHandler() at /var/www/app/setup/src/Magento/Setup/Model/Installer.php:861
 Magento\Setup\Model\Installer->handleDBSchemaData() at /var/www/app/setup/src/Magento/Setup/Model/Installer.php:775
 Magento\Setup\Model\Installer->installSchema() at /var/www/app/setup/src/Magento/Setup/Console/Command/DbSchemaUpgradeCommand.php:69
 Magento\Setup\Console\Command\DbSchemaUpgradeCommand->execute() at /var/www/app/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at /var/www/app/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at /var/www/app/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at /var/www/app/vendor/magento/framework/Console/Cli.php:102
 Magento\Framework\Console\Cli->doRun() at /var/www/app/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at /var/www/app/bin/magento:23
```

**Comment by Bang512** [comment]
Hi @andyexeter,

Sorry for the delay.
Please run the command line `php bin/magento setup:upgrade` to upgrade your database in your store.
Thanks!

**Comment by andyexeter** [comment]
Hi @Bang512,

Yes, this error happens when running `php bin/magento setup:upgrade`. 

The `setup:upgrade` is a multi-step command, with one of the steps being `setup:db-schema:upgrade` which is when this error occurs.

**Comment by AleksandrPlotnikov** [comment]
Any comment on this, I have the same issue.

**Comment by Bang512** [comment]
Hi @andyexeter,

Sorry for the delay, We just come back after the Lunar New Year Vacation.
Please try to run the command `rm -rf var/cache/ generated/` in the Magento root folder, then re-upgrade your database again.
Thanks

**Comment by Bang512** [comment]
Closed due no response within 7 days.

**Comment by sfresta** [comment]
I have the same issue. How to solve?

**Comment by Bang512** [comment]
Hi @sfresta,

Please download and replace the attached files:
 - <mageplaza_blog_path>/Setup/InstallData.php
 - <mageplaza_blog_path>/Setup/InstallSchema.php
 - <mageplaza_blog_path>/Setup/UpgradeData.php
 - <mageplaza_blog_path>/Setup/UpgradeSchema.php
[Setup.zip](https://github.com/mageplaza/magento-2-blog/files/4234932/Setup.zip)

Then, please re-upgrade your store and let me know the result.
Thanks!

**Comment by sfresta** [comment]
Hello Bang512,
now I receive the following error while upgrade:

Module 'Mageplaza_Core':
Module 'Mageplaza_Blog':
Installing schema... PHP Fatal error:  Uncaught Error: Call to a member function get() on null in /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:144
Stack trace:
#0 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(230): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'context', 'Mageplaza\\Blog\\...')
#1 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Mageplaza\\Blog\\...', Array, Array)
#2 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Mageplaza\\Blog\\...', Array, Array)
#3 /var/www/mysite/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Mageplaza\\Blo in /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php on line 144

Fatal error: Uncaught Error: Call to a member function get() on null in /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:144
Stack trace:
#0 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(230): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'context', 'Mageplaza\\Blog\\...')
#1 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Mageplaza\\Blog\\...', Array, Array)
#2 /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Mageplaza\\Blog\\...', Array, Array)
#3 /var/www/mysite/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Mageplaza\\Blo in /var/www/mysite/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php on line 144

**Comment by Bang512** [comment]
Hi @sfresta,

Please run the command line in the Magento installation folder:
`sudo rm -rf var/cache generated/`
before re-upgrade your store: `php bin/magento setup:upgrade`

**Comment by sfresta** [comment]
Same error man.. my Magento is v2.2.8 CE, if can help

**Comment by Bang512** [comment]
Hi @sfresta,

Please try this instead:
[Setup.zip](https://github.com/mageplaza/magento-2-blog/files/4235228/Setup.zip)

the path is the same in the previous message.
Thanks!

**Comment by sfresta** [comment]
[magento-2-blog-extension.zip](https://github.com/mageplaza/magento-2-blog/files/4235249/magento-2-blog-extension.zip)

I replaced vendor/mageplaza/magento-2-blog-extension/ with the one attached (found on net) and now the upgrade works

**Comment by Bang512** [comment]
Hi @sfresta,

Have you tried my solution? I recommend you should use our solution.
If you use the solution from the third party, I'm not sure you will get the best support on the feature with that fix.

**Comment by AleksandrPlotnikov** [comment]
after debugging code I solved this by removing etc/db_schema.xml and etc/db_schema_whitelist.xml, since we have install/upgrade scripts. (this is old approach but works)

**Comment by Bang512** [comment]
Hi @AleksandrPlotnikov,

With the Magento version lower v2.3, It will not  run the script in two files:
 - db_schema.xml
 - db_schema_whitelist.json
So, if you remove these files, it will not solve with the Magento version lower 2.3.0.
Thanks!

**Comment by asmithraj** [comment]
> Hi @sfresta,
> 
> Please try this instead:
> [Setup.zip](https://github.com/mageplaza/magento-2-blog/files/4235228/Setup.zip)
> 
> the path is the same in the previous message.
> Thanks!

This worked for me, thanks!!

**Comment by asmithraj** [comment]
> Hi @sfresta,
> 
> Please try this instead:
> [Setup.zip](https://github.com/mageplaza/magento-2-blog/files/4235228/Setup.zip)
> 
> the path is the same in the previous message.
> Thanks!

This worked for me. However i had to remove setAreaCode method from upgradeData.php and only keep it to installData.php to resolve another issue:  "Area code is already set" while running upgrade script. 

Thank you!!!

**Comment by sanaeefar-saeed** [comment]
> [magento-2-blog-extension.zip](https://github.com/mageplaza/magento-2-blog/files/4235249/magento-2-blog-extension.zip)
> 
> I replaced vendor/mageplaza/magento-2-blog-extension/ with the one attached (found on net) and now the upgrade works

worked for me
